### ISSUE TITLE
Call `PyType_Modified()` when an object's `__class__` is assigned to

### ISSUE THREAD

**Issue by swtaarrs** [python/cpython#103096]
# Feature or enhancement

To help type-specializing JITs, [`object_set_class()`](https://github.com/python/cpython/blob/ba65a065cf07a7a9f53be61057a090f7311a5ad7/Objects/typeobject.c#L5210-L5211) should call `PyType_Modified()` on the *old* type of the instance upon successful assignment. This operation doesn't modify the type in question, but it does have similar effects on the invariants needed by optimizing compilers.

# Pitch

The Cinder JIT supports generating type-specialized code. To avoid having to re-check an object's type every time it is used (after the initial guard), we hooked [`object_set_class()`](https://github.com/facebookincubator/cinder/blob/ca8f4a5aae161d3ff18cf395c62f894211f9cac4/Objects/typeobject.c#L5247) to notify the JIT when an object's `__class__` is assigned to. We then invalidate any JIT-compiled code that is specialized based on the old type.

The prototype hook linked above takes both the old and new types as arguments; this was intended to let us avoid invalidating the code if the new type also fits the constraints needed by the compiled code. I ended up not following through on this plan since it would mean extra complexity and bookkeeping overhead to help in a situation that should be vanishingly rare.

This should be a one-line code change (excluding tests): I'll insert a call to `PyType_Modified(oldto)` right after the successful assignment [here](https://github.com/python/cpython/blob/ba65a065cf07a7a9f53be61057a090f7311a5ad7/Objects/typeobject.c#L5307).

# Previous discussion

@carljm and @markshannon have discussed this offline.

**Comment by swtaarrs** [comment]
Sorry, I filed this issue a bit too soon. I just realized that a single call to `PyType_Modified(oldto)` won't give us everything that we need. We don't need both the old type and the new type like my prototype hook in Cinder, but we do need the old type and some indicator that `__class__` assignment has happened rather than an actual type modification. We don't invalidate JIT-compiled code on unrelated modifications to the type(s) in question (we can't afford to for performance reasons). We *do* invalidate all code related to a specific type when a `__class__` assignment away from that type happens, though, so the JIT needs some way of distinguishing between the two situations.

I'm exploring other options at the moment and will report back later this week.
